man/list-apt-installs.1
.\" Manpage for list-apt-installs.
.\" Contact zcotcaudle@gmail.com to correct errors or typos.
.TH man 1 "06 May 2018" "0.6" "list-apt-installs man page"
.SH NAME
list-apt-installs \- output list of apt/apt-get USER installed packages from apt log(s).
.SH SYNOPSIS
.B list-apt-installs
[\fI\,OPTIONS\/\fR] [\fI\,-o|--output\/\fR[\fI\,=\/\fR] \fI\,</path/and/or/file/name>\/\fR]
.SH DESCRIPTION
list-apt-installs is for listing successful log entries derived from using 'apt install' or 'apt-get install' terminal commands. The default output is formatted for use with the restore function in Linux Mint Backup
tool, or installing with `apt\-get dselect\-upgrade` method. Or output can be formatted as a 'apt install' reinstall script which will need \fBchmod +x\fR to run.
.SH OPTIONS
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,FILE\/\fR
Output to FILE instead of standard output
.TP
\fB\-s\fR, \fB\-\-script\fR
Output as `apt install` script.
.TP
\-?, \fB\-\-help\fR
Give this help list
.TP
\fB\-\-usage\fR
Give a short usage message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print program version
.PP
Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.
.SH SEE ALSO
\fBapt-get\fR(8),\fBdselect\fR(1)
.SH AUTHOR
Scott Caudle (zcotcaudle@gmail.com)


CHANGELOG
list-apt-installs (0.6) Linux Mint 19; urgency=medium

  * check if package is still installed, will not include removed/purged packages
  * tweaks and bug fixes for parsing command line flags

-- Scott Caudle <zcotcaudle@gmail.com>  Wed, 1 Aug 2018 22:18:47 -0400

list-apt-installs (0.5) Linux Mint 18.x; urgency=medium

  * Implement arg --script to output in 'apt install' format.
  * Implement arg --output file capability, optional from piping output.

-- Scott Caudle <zcotcaudle@gmail.com>  Mon, 18 Jun 2018 14:20:21 -0400

list-apt-installs (0.4) Linux Mint 18.x; urgency=medium

  * Add argc/argv parsing. Implement --help and --version.
  * Expanding on README.md

-- Scott Caudle <zcotcaudle@gmail.com>  Sun, 17 Jun 2018 16:09:36 -0400

list-apt-installs (0.3) Linux Mint 18.x; urgency=medium

  * New initial release.

-- Scott Caudle <zcotcaudle@gmail.com>  Wed, 30 May 2018 23:59:13 -0400


source/list-apt-installs.c
//
// Copyright 2018 Scott Caudle ( zcotcaudle@gmail.com ) - zcot ( https://github.com/zcot )
// This program is released under GPL-3.0 or any later version.
//

/*                LICENSE NOTICE:
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/


/*                     list-apt-installs:
*
*   The program finds apt and apt-get install commands in apt logs that were typed in terminal.
*
*   -traversing /var/log/apt/history.log and archived history logs.
*
*   It produces output that is suitable with the restore functionality from Mint's backup tool.
*
*   You can produce a ".list" file by simply piping the output to a file.
*
*   See man list-apt-installs or lai --help
*
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <argp.h>

const char *argp_program_version = "0.6";
const char *argp_program_bug_address = "<zcotcaudle@gmail.com>";
static char doc[] = "list-apt-installs -- make a list of previous apt and apt-get installs that were made.\
\vDefault output is formatted for use with restore function in Linux Mint Backup tool, \
or installing with `apt-get dselect-upgrade` method.";
static char args_doc[] = "[-o|--output[=] </path/and/or/file/name>]";

// argc/argv flags
static struct argp_option options[] = {
  {"script",  's', 0,       0, "Output as `apt install` script." },
  {"output",   'o', "FILE",  0, "Output to FILE instead of standard output" },
  { 0 }
};

// flags structure
struct arguments
{
  char* arg1;          // for file output
  char** strings;      // for file output
  char* output_file;   // file arg to --output
  int script_building; // boolean --script
};

// parse each flag
static error_t parse_opt(int key, char* arg, struct argp_state* state)
{
  struct arguments* arguments = state->input;

  switch (key)
    {
    case 's':
      arguments->script_building = 1;
      break;
    case 'o':
      arguments->output_file = arg;
      break;

    case ARGP_KEY_ARG:
      arguments->arg1 = arg;
      arguments->strings = &state->argv[state->next];
      state->next = state->argc;
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

// argp parser
static struct argp argp = {options, parse_opt, args_doc, doc};

char g_ending[12];
FILE* g_fout = NULL;


// a boolean type of package checking utility
int package_exists(char* name)
{
    FILE* pdpkgquery = NULL;
    char buf[4096];
    char query[4096];
    char pkgname[255];

    //strip specified version if any was provided. for example: apt install hello=2.10-3build1
    char* equal = strchr(name, '=');
    if(equal)
    {
        strncpy(pkgname, name, equal-name);
        pkgname[equal-name] = '\0';
        snprintf(query, 4096, "dpkg-query --show --showformat='${db:Status-Status}\n' %s 2>/dev/null", pkgname);
    }
    else
    snprintf(query, 4096, "dpkg-query --show --showformat='${db:Status-Status}\n' %s 2>/dev/null", name);

    pdpkgquery = popen(query, "r");
    if(!pdpkgquery)
    {
        fprintf(stderr, "lai-fail: popen: %s\n", strerror(errno));
        return 0;
    }
    if(fgets(buf, 4096, pdpkgquery))
        //if(strncmp("installed", buf, 9) == 0)
        // successful output should only be "installed" or "not-installed"
        // other than erroneous condition piped out the stderr which we dont care
        if(buf[0] == 'i')
        {
            pclose(pdpkgquery);
            return 1;
        }

    pclose(pdpkgquery);
    return 0;
}



int parse_line_and_output(char* line)
{
    // line will only be a "Commandline: " type passed to this function
    // and here 'line' starts after the "Commandline: " part
    //example: "Commandline: apt-get install whatever"
    const char aptget[] =  "apt-get "; //length is 8, we start checking at 9(index 8)
    const char aptmint[] = "/usr/bin/apt install "; //length 21, start at index 21
    const char apt[] =     "apt install "; //length 12

    //break up any multi-package calls...
    // ex. apt install screenfetch lolcat timg conky-manager2
    //and check for and ignore extra command flags (anything started with -)
    const char s[2] = " ";
    char* tok;

    //truncate newline
    char* newline = strchr(line, '\n');
    if(newline) *newline = '\0';

    //copy the package(s) based on which command was used
    //apt-get can use a flag in front of 'install' and after so check for 'apt-get' then find 'install' anywhere
    if(strncmp(aptget, line, 8) == 0)
    {
        if(strstr(&line[7], " install "))
        {
            if(strstr(&line[7], " --force-yes "))//if force-yes then bail out
                return 0;

            tok = strtok(&line[8], s);
        }
        else return 0;
    }
    else if(strncmp(apt, line, 12) == 0)
    {
        tok = strtok(&line[12], s);
    }
    else if(strncmp(aptmint, line, 21) == 0)
    {
        tok = strtok(&line[21], s);
    }
    else return 0;

    while(tok)
    {
        //skip the token if it's a command flag or switch
        if(tok[0] != '-')
            if(package_exists(tok))
            fprintf(g_fout, "%s %s\n", tok, g_ending);

        tok = strtok(NULL, s);
    }
    return 0;
}



// a boolean type of file checking utility
int file_exists(char* name)
{
    FILE* f = fopen(name, "r");

    if(f)
    {
        fclose(f);
        return 1;
    }

    return 0;
}



// loop through each log file and parse out the data we are interested in
int main(int argc, char* argv[])
{
    const char prefix[] = "zgrep Commandline: ";
    char* cmd;
    FILE* in;
    char buf[4096];
    char name[128];
    memset(name, '\0', sizeof(name));
    memset(g_ending, '\0', sizeof(g_ending));

    struct arguments arguments;
    // default args
    arguments.script_building = 0; // not building script
    arguments.output_file = "-"; // not using file output

    //take args and adjust
    if(argp_parse(&argp, argc, argv, 0, 0, &arguments))
        exit (0);

    cmd = malloc(sizeof(prefix) + 128 + 1);
    if(!cmd)
    {
        fprintf(stderr, "%s: malloc: %s\n", argv[0], strerror(errno));
        return 1;
    }

    
    //start here...
    strcpy(name, "/var/log/apt/history.log");

    //using stdout or file?
    if(strcmp(arguments.output_file, "-"))
    {
        g_fout = fopen(arguments.output_file, "w");
    }
    else
        g_fout = stdout;

    //building output as script or not?
    if(arguments.script_building)
    {
        fprintf(g_fout, "#/bin/bash\napt install \\\n");
        strcpy(g_ending, "\\");
    }
    else
        strcpy(g_ending, "install");

    int n = 1;
    while(file_exists(name))
    {
        //set grep command and file name
        sprintf(cmd, "%s%s", prefix, name);

        //will only return "Commandline: " lines
        in = popen(cmd, "r");
        if(!in)
        {
            fprintf(stderr, "%s: popen: %s\n", argv[0], strerror(errno));
            return 1;
        }

        while(fgets(buf, 4096, in))
        {
            //parse for data and do output
            //send line starting after "Commandline: " part
            parse_line_and_output(&buf[13]);
        }
        pclose(in);

        // seek the numbered archives
        sprintf(name, "%s%i%s", "/var/log/apt/history.log.", n, ".gz");
        n++;
    }

    fclose(g_fout);// closing stdout when file is not used?
    free(cmd);

    return 0;
}



